sui_core/src/authority/authority_store.rs

```diff
@@ -303,14 +307,43 @@ impl<
             .collect())
     }
 
+    // Returns true if the most recent version of the object does not correspond to a deleted
+    // version.
+    fn object_is_live(&self, object_id: &ObjectID) -> Result<bool, SuiError> {
+        let obj = self
+            .parent_sync
+            .iter()
+            .skip_prior_to(&(*object_id, SequenceNumber::MAX, ObjectDigest::MAX))?
+            .next();
+
+        Ok(obj.map(|(k, _)| k.2.is_live()).unwrap_or(false))
+    }
+
     /// Read an object and return it, or Err(ObjectNotFound) if the object was not found.
     pub fn get_object(&self, object_id: &ObjectID) -> Result<Option<Object>, SuiError> {
-        self.objects.get(object_id).map_err(|e| e.into())
+        if !self.object_is_live(object_id)? {
+            Ok(None)
+        } else {
+            let obj = self
+                .objects
+                .iter()
+                .skip_prior_to(&ObjectKey::max_for_id(object_id))?
+                .next();
+
+            match obj {
+                Some((ObjectKey(obj_id, _), obj)) if obj_id == *object_id => Ok(Some(obj)),
+                _ => Ok(None),
```

Comment:
Added an error log.

I believe I fixed the race by checking parent_sync after loading the object. This is still a race, but it's a race that could happen even the two reads here were atomic - the object could be deleted right before the first read.

I would like to leave the `object_id` signature change for another PR. It affects `check_gas` which I don't fully understand. Either the existing `check_gas` method is broken (because it might consult a newer version of the gas object which breaks idempotency), or changing it to use a fixed version would be broken (because it might consult an old version). I'd like to understand that issue better before proceeding.